titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/EntityHolder.java

```diff
@@ -76,34 +88,26 @@ public String getId() {
     }
 
     public EntityHolder addChild(EntityHolder child) {
-        SortedSet<EntityHolder> newChildren = new TreeSet<>(Comparator.comparing(EntityHolder::getId));
-        newChildren.addAll(children);
-
-        // Set will not be updated if the child already is there, so explicitly remove it.
-        newChildren.remove(child);
-        newChildren.add(child);
-
-        return new EntityHolder(id, entity, newChildren, attributes);
+        Map<String, EntityHolder> newChildrenById = new HashMap<>(childrenById);
+        newChildrenById.put(child.getId(), child);
+        return new EntityHolder(id, entity, newChildrenById, attributes);
     }
 
     public Pair<EntityHolder, Optional<EntityHolder>> removeChild(String id) {
-        return children.stream().filter(c -> c.getId().equals(id)).findFirst().map(removedChild -> {
-                    SortedSet<EntityHolder> filteredChildren = new TreeSet<>(Comparator.comparing(EntityHolder::getId));
-                    children.forEach(c -> {
-                        if (!c.getId().equals(id)) {
-                            filteredChildren.add(c);
-                        }
-                    });
-                    EntityHolder newRoot = new EntityHolder(this.id, this.entity, filteredChildren, this.attributes);
-                    return Pair.of(newRoot, Optional.of(removedChild));
-                }
-        ).orElseGet(() -> Pair.of(this, Optional.empty()));
+        if (!childrenById.containsKey(id)) {
+            return Pair.of(this, Optional.empty());
+        }
+        Map<String, EntityHolder> newChildrenById = new HashMap<>(childrenById);
+        EntityHolder removedChild = newChildrenById.remove(id);
+
+        EntityHolder newRoot = new EntityHolder(this.id, this.entity, newChildrenById, this.attributes);
+        return Pair.of(newRoot, Optional.of(removedChild));
```

Comment:
`Optional.ofNullable`

Reply:
We check that the child exists, so the removedChild is always non-null.